File: online-judge/4th-semester/exam3/11402_AhoyPirates.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

const int MAXN = 1030000;

vector<pair<int, int>> tree(MAXN * 2);
vector<string> pend(MAXN * 2, "");

pair<int, int> invValue()
{
    return make_pair(0, 0);
}

pair<int, int> createValue(int val)
{
    if(val == '0') return make_pair(0, 1);
    else return make_pair(1, 0);
}

pair<int, int> combine(pair<int, int> p1, pair<int, int> p2)
{
    return make_pair(p1.first + p2.first, p1.second + p2.second);
}

pair<int, int> change(int v, string &str)
{
    int tmp;
    pair<int, int> ans = tree[v];
    for(int i = 0; i < str.size(); i++)
    {
        if (str[i] == 'F')
        {
            ans.first = ans.first + ans.second;
            ans.second = 0;
        }
        else if (str[i] == 'E')
        {
            ans.second = ans.first + ans.second;
            ans.first = 0;
        }
        else if (str[i] == 'I')
        {
            tmp = ans.first;
            ans.first = ans.second;
            ans.second = ans.first;
        }
    }
    return ans;
}

void push(int v, int v1, int v2)
{
    tree[v1] = change(v1, pend[v]);
    pend[v1] += pend[v];
    tree[v2] = change(v2, pend[v]);
    pend[v2] += pend[v];
    pend[v] = "";
}

void build(string &arr, int v, int l, int r)
{
    if(l == r)
    {
        tree[v] = createValue(arr[l]);
    }
    else
    {
        int mid = l + ((r - l) >> 1);
        build(arr, v + 1, l, mid);
        build(arr, v + 2 * (mid - l + 1), mid + 1, r);
        tree[v] = combine(tree[v + 1], tree[v + 2 * (mid - l + 1)]);
    }
}

pair<int, int> query(int v, int L, int R, int l, int r)
{
    pair<int, int> ans;
    if(l > r)
    {
        ans = invValue();
    }
    else if(l == L && r == R) 
    {
        ans = tree[v];
    }
    else
    {
        int m = L + ((R - L) >> 1);
        push(v, v + 1, v + 2 * (m - L + 1));
        ans = combine(query(v + 1, L, m, l, min(r, m)), 
                      query(v + 2 * (m - L + 1), m + 1, R, max(l, m + 1), r));

    }
    return ans;
}

void update(int v, int L, int R, int l, int r, string &h)
{
    if(l <= r)
    {
        if(l == L && r == R)
        {
            tree[v] = change(v, h);
            pend[v] += h;
        }
        else
        {
            int mid = L + ((R - L) >> 1);
            push(v, v + 1, v + 2 * (mid - L + 1));
            update(v + 1, L, mid, l, min(r, mid), h);
            update(v + 2 * (mid - L + 1), mid + 1, R, max(l, mid + 1), r, h);
            tree[v] = combine(tree[v + 1], tree[v + 2 * (mid - L + 1)]);
        }
    }
}


int main()
{
    std::ios_base::sync_with_stdio(false);
    int cases, mPairs, mult, last, queries;
    int left, right, ans;
    int countGod, count;
    count = 1;
    string op;
    cin >> cases;
    while(cases--)
    {
        string arr = "";
        cin >> mPairs;
        while(mPairs--)
        {
            cin >> mult;
            cin >> op;
            for(int i = 0; i < mult; i++)
            {
                arr += op;
            }
        }
        last = arr.size() - 1;
        build(arr, 0, 0, last);
        countGod = 1;
        cin >> queries;
        cout << "Case " << count++ << ":\n";
        while(queries--)
        {
            cin >> op >> left >> right;
            if(op == "S")
            {
                ans = query(0, 0, last, left, right).first;
                cout << "Q" << countGod++ << ": " << ans << "\n";
            }
            else
            {
                update(0, 0, last, left, right, op);
            }
        }
    }
    return 0;
}

File: online-judge/4th-semester/exam3/11402_AhoyPirates.py
from sys import stdin, stdout

MAXN = 1030000
tree = [None for _ in range(MAXN * 2)]
pend = [""] * (MAXN * 2)

def reset():
    global pend
    pend = [""] * (MAXN * 2)

def invValue():
    return (0, 0)

def createValue(val):
    #(Buccaneer, Barbary)
    if val == '0':
        return (0, 1)
    else:
        return (1, 0)

def combine(p1, p2):
    return (p1[0] + p2[0], p1[1] + p2[1])

def push(v, v1, v2):
    l = pend[v]
    tree[v1] = change(v1, l)
    pend[v1] += l
    tree[v2] = change(v2, l)
    pend[v2] += l
    pend[v] = ""

    # if pend[v] == "F":
    #     tree[v1] = (tree[v1][0] + tree[v1][1], 0) 
    #     pend[v1] = "F"
    #     tree[v2] = (tree[v2][0] + tree[v2][1], 0) 
    #     pend[v2] = "F"

    # elif pend[v] == "E":
    #     tree[v1] = (0, tree[v1][0] + tree[v1][1]) 
    #     pend[v1] = "E"
    #     tree[v2] = (0, tree[v2][0] + tree[v2][1]) 
    #     pend[v2] = "E"

    # elif pend[v] == "I":
    #     tree[v1] = (tree[v1][1], tree[v1][0]) 
    #     pend[v1] = "I"
    #     tree[v2] = (tree[v2][1], tree[v2][0]) 
    #     pend[v2] = "I"
    # pend[v] = ""

    

def change(v, string):
    ans = tree[v]
    for l in string:
        if l == "F":
            ans = (ans[0] + ans[1], 0) 
        elif l == "E":
            ans = (0, ans[0] + ans[1]) 
        elif l == "I":
            ans = (ans[1], ans[0]) 
    return ans

#build the segment tree
def build(a, v, l, r):
    if l == r: tree[v] = createValue(a[l])
    else:
        m = l + ((r - l) >> 1)
        build(a, v + 1, l, m)
        build(a, v + 2 * (m - l + 1), m + 1, r)
        tree[v] = combine(tree[v + 1], tree[v + 2 * (m - l + 1)])

#query
def query(v, L, R, l, r):
    ans = None
    if l > r: ans = invValue()
    elif l == L and r == R: ans = tree[v]
    else:
        m = L + ((R - L) >> 1)
        push(v, v + 1, v + 2 * (m - L + 1))
        ans = combine(query(v + 1, L, m, l, min(r, m)), 
                      query(v + 2 * (m - L + 1), m + 1, R, max(l, m + 1), r))
        #print(f"combined {l} {r} = {ans}")
    return ans

def update(v, L, R, l, r, h : str):
    if l <= r:
        if l == L and r == R: 
            tree[v] = change(v, h)
            pend[v] += h
        else:
            m = L + ((R - L) >> 1)
            push(v, v + 1, v + 2 * (m - L + 1))
            update(v + 1, L, m, l, min(r, m), h)
            update(v + 2 * (m - L + 1), m + 1, R, max(l, m + 1), r, h)
            tree[v] = combine(tree[v + 1], tree[v + 2 * (m - L + 1)])
            #print(f"combined {l} {r} = {tree[v]}")
            


def main():
    cases = int(stdin.readline())
    for i in range(1, cases + 1):
        if i != 1:
            reset()
        string = ""
        mPairs = int(stdin.readline())
        for _ in range(mPairs):
            mult = int(stdin.readline())
            subString = stdin.readline().rstrip()
            string += (subString * mult)
        last = len(string) - 1
        #print(string)
        build(string, 0, 0, last)
        queries = int(stdin.readline())
        count = 1
        stdout.write(f"Case {i}:\n")
        for _ in range(queries):
            info = stdin.readline().split()
            op = info[0]
            left = int(info[1])
            right = int(info[2])
            if op == "S":
                #print(op, left, right)
                ans = query(0, 0, last, left, right)[0]
                stdout.write(f"Q{count}: {ans}\n")
                count += 1
            else:
                #print(op, left, right)
                update(0, 0, last, left, right, op)
            # for j in range(last * 2 + 2):
            #     print(tree[j], end=" ")
            # print()


main()